server/remote_cache/action_cache_server/action_cache_server.go

```diff
@@ -242,3 +244,58 @@ func (s *ActionCacheServer) UpdateActionResult(ctx context.Context, req *repb.Up
 	}
 	return req.ActionResult, nil
 }
+
+// Inlines the contents of output files requested to be inlined as long as the
+// total size of the ActionResult is below maxResultSize.
+func (s *ActionCacheServer) maybeInlineOutputFiles(ctx context.Context, req *repb.GetActionResultRequest, ar *repb.ActionResult, maxResultSize int) {
+	if ar == nil || len(req.InlineOutputFiles) == 0 {
+		return
+	}
+	requestedFiles := make(map[string]struct{}, len(req.InlineOutputFiles))
+	for _, f := range req.InlineOutputFiles {
+		requestedFiles[f] = struct{}{}
+	}
+
+	budget := int64(max(0, maxResultSize-proto.Size(ar)))
+	var actuallyInlinedFiles []*repb.OutputFile
+	for _, f := range ar.OutputFiles {
+		if _, ok := requestedFiles[f.Path]; !ok {
+			continue
+		}
+		contentsSize := f.GetDigest().GetSizeBytes()
+		if contentsSize == 0 {
+			// Empty files don't need to be inlined as they can be recognized
+			// by their size.
+			continue
+		}
+		// An additional "contents" field requires 1 byte for the tag field
+		// (5:LEN), the bytes for the varint encoding of the length of the
+		// contents and the contents themselves.
+		totalSize := 1 + int64(binary.Size(uint64(contentsSize))) + contentsSize
+		if budget < totalSize {
+			continue
+		}
+		budget -= totalSize
+		actuallyInlinedFiles = append(actuallyInlinedFiles, f)
+	}
+
+	if len(actuallyInlinedFiles) == 0 {
+		return
+	}
+
+	g, gCtx := errgroup.WithContext(ctx)
+	for _, f := range actuallyInlinedFiles {
+		fc := f
+		g.Go(func() error {
+			rn := digest.NewResourceName(fc.GetDigest(), req.GetInstanceName(), rspb.CacheType_CAS, req.GetDigestFunction()).ToProto()
+			blob, err := s.cache.Get(gCtx, rn)
```

Comment:
Does this update the usual download metrics?

Reply:
Good catch - hits/misses are tracked in the API layer rather than the storage layer (cache) so I think you'd need to construct a HitTracker here and track any hits/misses.

One interesting thing is that each hit tracked will appear as a separate row in the cache requests card, even though there is only a single client "request," though this seems OK for now. In the future, it'd be neat to somehow attach the AC request info to these rows, so that we can show them in the UI as "inlined" or something.